pkg/api/networkservice/mechanisms/vfio/helpers.go

```diff
@@ -17,51 +18,124 @@
-	GetNetNSInode() string
-	GetParameters() map[string]string
-	GetPCIAddress() string
+// Mechanism is a vfio mechanism helper
+type Mechanism struct {
+	*networkservice.Mechanism
 }
 
-type mechanism struct {
-	*networkservice.Mechanism
+// New returns *networkservice.Mechanism of type vfio with the given cgroupDir
+func New(cgroupDir string) *networkservice.Mechanism {
+	return &networkservice.Mechanism{
+		Cls:  cls.LOCAL,
+		Type: MECHANISM,
+		Parameters: map[string]string{
+			CgroupDirKey: cgroupDir,
+		},
+	}
 }
 
 // ToMechanism converts unified mechanism to helper
-func ToMechanism(m *networkservice.Mechanism) Mechanism {
+func ToMechanism(m *networkservice.Mechanism) *Mechanism {
 	if m.GetType() == MECHANISM {
-		return &mechanism{
+		return &Mechanism{
 			m,
 		}
 	}
 	return nil
 }
 
-// GetParameters returns map of mechanism parameters
-func (m *mechanism) GetParameters() map[string]string {
-	if m == nil {
-		return nil
+// GetParameters returns the map of all parameters to the mechanism
+func (m *Mechanism) GetParameters() map[string]string {
+	if m.Parameters == nil {
+		m.Parameters = map[string]string{}
 	}
 	return m.Parameters
 }
 
-// GetParameters returns network namespace Inode
-func (m *mechanism) GetNetNSInode() string {
-	if m == nil || m.GetParameters() == nil {
-		return ""
-	}
-	return m.GetParameters()[common.NetNSInodeKey]
+// GetCgroupDir returns client on host cgroup directory
+func (m *Mechanism) GetCgroupDir() string {
+	return m.GetParameters()[CgroupDirKey]
+}
+
+// SetCgroupDir sets client on host cgroup directory
+func (m *Mechanism) SetCgroupDir(cgroupDir string) {
+	m.Parameters[CgroupDirKey] = cgroupDir
+}
+
+// GetIommuGroup returns IOMMU group id
+func (m *Mechanism) GetIommuGroup() uint {
+	return atou(m.Parameters[IommuGroupKey])
+}
+
+// SetIommuGroup sets IOMMU group id
+func (m *Mechanism) SetIommuGroup(iommuGroup uint) {
+	m.Parameters[IommuGroupKey] = utoa(iommuGroup)
+}
+
+// GetPCIAddress returns PCI address
+func (m *Mechanism) GetPCIAddress() string {
+	return m.GetParameters()[PCIAddressKey]
+}
+
+// SetPCIAddress sets PCI address
+func (m *Mechanism) SetPCIAddress(pciAddress string) {
+	m.Parameters[PCIAddressKey] = pciAddress
+}
+
+// GetVfioMajor returns /dev/vfio major number
+func (m *Mechanism) GetVfioMajor() uint32 {
+	return uint32(atou(m.Parameters[VfioMajorKey]))
 }
 
-// GetPCIAddress returns PCI address of the VF device
-func (m *mechanism) GetPCIAddress() string {
-	if m == nil || m.GetParameters() == nil {
-		return ""
+// SetVfioMajor sets /dev/vfio major number
+func (m *Mechanism) SetVfioMajor(vfioMajor uint32) {
+	m.Parameters[VfioMajorKey] = utoa(uint(vfioMajor))
+}
+
+// GetVfioMinor returns /dev/vfio minor number
+func (m *Mechanism) GetVfioMinor() uint32 {
+	return uint32(atou(m.Parameters[VfioMinorKey]))
+}
+
+// SetVfioMinor sets /dev/vfio minor number
+func (m *Mechanism) SetVfioMinor(vfioMinor uint32) {
+	m.Parameters[VfioMinorKey] = utoa(uint(vfioMinor))
```

Comment:
Won't this panic if m == nil or Parameters == nil?

Reply:
Fixed, now GetParameters() returns empty map for m == nil.